### PR TITLE
Use Poetry for package management

### LINKED ISSUE #691: Improve way of managing python dependencies
## Description

As a developer,
I need a more flexible and easy to manage way of managing python dependencies,
so that I can update in a more reliable way the packages that are being flagged as security risks and ease development to coder contributors.

A few options are:
- https://github.com/jazzband/pip-tools
- https://python-poetry.org/
- feel free to suggest more...

Since the beginning of the project I have been installing and just do `pip freeze > requirements.txt` this can make me install unrequired packages because of something I might install into the environment which is not necessary. It feels clumsy, I know we can have a better way to do this.

### Update [Required]
- [ ] Adopt a way of managing python dependencies
- [ ] Discuss with the community the tooling selected
- [ ] Document the selected way on README

## Definition of Done
- [ ] All of the required items are completed.
- [x] Approval by 1 mentor.

## Estimation
5 hours

### PR DESCRIPTION
**Pull request by bartekpacia**
### Description
Migrated to [Poetry](https://python-poetry.org) as our default dependency management tool. It's simple, yet powerful.

- added `pyproject.toml` and `pyproject.loc`
- `requirements.txt` were regenerated using [poetry export](`https://python-poetry.org/docs/cli/#export) - it will be used only to speed up dependency resolution during CI builds ([discussion here](https://github.com/anitab-org/mentorship-backend/pull/715#discussion_r466545400))

fixes #691 
furthermore, #715 can be closed. #715 was the first version of the PR, but it didn't age well (diff too big) so I decided to start fresh with this one.

### Type of Change:

- Code
- Documentation

**Code/Quality Assurance Only**
- New feature (non-breaking change which adds functionality pre-approved by mentors)

### How Has This Been Tested?

I ran the test suite, all tests passed :)

### Checklist:

- [x] My PR follows the style guidelines of this project
- [x] I have performed a self-review of my own code or materials
- [x] I have commented my code or provided relevant documentation, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [ ] Any dependent changes have been merged
- [ ] Update Postman API at /docs folder
- [ ] Update Swagger documentation and the exported file at /docs folder
- [ ] ~Update requirements.txt~ Not anymore! Updated `pyproject.toml` haha ;)

**Code/Quality Assurance Only**

- [x] My changes generate no new warnings 
- [x] My PR currently breaks something (fix or feature that would cause existing functionality to not work as expected)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been published in downstream modules

### CODE REVIEW

**Comment by bartekpacia** [comment]
Ooh, is it __that__ bad that (if I can read) the code coverage dropped by 0.10%?

File: README.md [hunk 1]
```diff
@@ -34,19 +33,23 @@ To setup the project locally read these wiki pages and follow the instructions:
  - [Fork, Clone and Remote](https://github.com/anitab-org/mentorship-backend/wiki/Fork%2C-Clone-%26-Remote)
  - [Export Environment Variables](docs/environment-variables.md)
 
-### Run app in Windows
+### Run app locally
 
-The project runs on Python 3.
+The project runs on Python 3.9. We use [Poetry](https://python-poetry.org/) to manage dependencies. It's powerful yet simple to use.
```
Comment (devkapilbansal) [inline README.md]: @bartekpacia we should test it for other python versions as well. Python 3.9 is launched recently and not all users have migrated to it

Comment (bartekpacia) [reply README.md]: I agree, however, we shouldn't hold onto the old versions as well.
I will test on 3.8.6, 3.7.9 and 3.6.12

**Comment by bartekpacia** [comment]
All tests pass on the following Python versions:
- 3.9
- 3.8.6
- 3.7.9
- 3.6.12

I suggest including a recommendation (e.g "We recommend using Python 3.9, but any Python >= 3.6 should work") in the  README.

File: README.md (hunk 1 above)
Comment (devkapilbansal) [reply README.md]: Yes, just test them from python>=3.6

Comment (bartekpacia) [reply README.md]: @devkapilbansal done :)

**Comment by devkapilbansal** [comment]
> All tests pass on the following Python versions:
> 
> * 3.9
> * 3.8.6
> * 3.7.9
> * 3.6.12
> 
> I suggest including a recommendation (e.g "We recommend using Python 3.9, but any Python >= 3.6 should work") in the README.

You can go like this, but first squash your commits in one. It's better to use one commit for one change when making pull requests. That helps us keep the history clean

**Comment by bartekpacia** [comment]
@devkapilbansal squashed :)

**Comment by bartekpacia** [comment]
@isabelcosta @devkapilbansal I'd love to hear what you think about this PR! :)

File: README.md [hunk 2]
```diff
@@ -34,29 +33,23 @@ To setup the project locally read these wiki pages and follow the instructions:
...
+Here are a few basic commands you'll use the most:
+- `poetry install` - installs the dependencies from `pyproject.toml` package
+- `poetry add <package>` - adds the `<package>` to `pyproject.toml` and installs it
+- `poetry remove <package>`- removes the `<package>` from `pyproject.toml` and uninstalls it
+- `poetry shell` - spawns a shell within the virtual environment. Equivalent of `virtualenv activate`, for those familiar with [`virtualenv`](https://pypi.org/project/virtualenv/)
+- `deactivate` - deactivates the virtual environemnt. Counterpart for `poetry shell`. Use this when you finish work.
```
Comment (isabelcosta) [inline README.md]: could you put these in a subsection about managing dependencies perhaps, it looks confusing being just together with the steps. being in a separate and clear section would be easier to skim through.

**Review by isabelcosta — COMMENTED** [review]
@bartekpacia I am getting this error while running your code here (from your repository), is there any missing package here?
image

Also, can you please fix the merge conflicts?

File: README.md (hunk 2 above)
Comment (bartekpacia) [reply README.md]: I redesigned this section a bit. Let me know what you think:)

**Comment by bartekpacia** [comment]
@isabelcosta That "module not found" error might've been caused by the lack of a valid `.env` file. I encountered it myself when I didn't have `.env` present in the project directory. As soon as I added it, it started to work fine (btw it's clearly explained in setup instructions in README that you _do need_ `.env` file, and apparently we both missed it)

The minimal `.env` looks like this:
```
FLASK_ENVIRONMENT_CONFIG = local
SECRET_KEY = <your-secret-key>
SECURITY_PASSWORD_SALT = <your-security-password-salt>
MAIL_DEFAULT_SENDER = <mail-default-sender>
MAIL_SERVER = <mail-server>
APP_MAIL_USERNAME = <app-mail-username>
APP_MAIL_PASSWORD = <app-mail-password>
MOCK_EMAIL = True
```

I think this PR is ready to be merged, so I'm asking for a review.

File: README.md [hunk 3]
```diff
@@ -213,6 +140,15 @@ Here are some links to documentation for this project:
 
 Understand more about our technical decisions made along with this project development in [Technical Decisions Wiki page](https://github.com/anitab-org/mentorship-backend/wiki/Technical-Decisions).
 
+### Poetry commands
+As said earlier, we use Poetry. It's simple yet powerful, but we know that many people use other package management tools.
```
Comment (isabelcosta) [inline README.md]:
```suggestion
As said earlier, we use [Poetry](https://python-poetry.org/).
```
Small suggestion, I removed the 2nd part of the sentence to make this intro to the commands shorter. With regards to documentation, it's important to be concise right? (that is why the run step was reduced again to 1 mini section that supports both windows and linux/mac run) So there's no need to leave opinionated text 🤔 It seems like filler.

File: README.md [hunk 4]
```diff
@@ -213,6 +140,15 @@ Here are some links to documentation for this project:
...
+To get you up to speed with Poetry, below we listed a few basic commands you'll use the most:
```
Comment (isabelcosta) [inline README.md]:
```suggestion
To get you up to speed, here's a few basic commands you'll use the most:
```

File: README.md (hunk 2 above)
Comment (isabelcosta) [reply README.md]: Looks good :) I just added a comment about "improving" the text that makes the intro to the poetry commands.

**Comment by isabelcosta** [comment]
> @isabelcosta That "module not found" error might've been caused by the lack of a valid `.env` file. I encountered it myself when I didn't have `.env` present in the project directory. As soon as I added it, it started to work fine (btw it's clearly explained in setup instructions in README that you _do need_ `.env` file, and apparently we both missed it)
> 
> The minimal `.env` looks like this:
> 
> ```
> FLASK_ENVIRONMENT_CONFIG = local
> SECRET_KEY = <your-secret-key>
> SECURITY_PASSWORD_SALT = <your-security-password-salt>
> MAIL_DEFAULT_SENDER = <mail-default-sender>
> MAIL_SERVER = <mail-server>
> APP_MAIL_USERNAME = <app-mail-username>
> APP_MAIL_PASSWORD = <app-mail-password>
> MOCK_EMAIL = True
> ```
> 
> I think this PR is ready to be merged, so I'm asking for a review.

I already have this file in the same project environment with your branch, and it still does not work. Are all dependencies correct? I tried doing `poetry install` and `poetry shell`, and it says there is nothing to install and update. I will look into this later today. @bartekpacia

File: README.md (hunk 3 above)
Comment (bartekpacia) [reply README.md]: You're right, that sentence doesn't add much value

**Comment by bartekpacia** [comment]
Hmm, this indeed sounds strange. I had no luck in trying to reproduce that bug on my side.

**Review by isabelcosta — COMMENTED** [review]
I see now @bartekpacia this works when I use local app environment:
```
export FLASK_ENVIRONMENT_CONFIG=local
```
In this case, a local sqlite database is used instead of connecting with a remote database.
On Heroku, and also on my machine environment, I am connecting to a remote Postgres database as a service called [ElephantSQL](https://www.elephantsql.com/).
```
export DB_TYPE='postgres'
export DB_USERNAME='xxxx'
export DB_PASSWORD='xxxxx'
export DB_ENDPOINT='drona.db.elephantsql.com:5432'
export DB_NAME='xxxx'
```
And for that case, I guess we need that missing package which is not being installed with this PR's package requirements.

Can you look into this? Perhaps all you need is adding this extra package 🤔 This would fail as it is now when deployed on heroku.

**Comment by bartekpacia** [comment]
@isabelcosta I've no experience with SQL databases except some simple SQLite and have no idea what package might be missing:/ Can you guide me a bit?

File: Dockerfile [hunk 5]
```diff
@@ -1,6 +1,6 @@
-FROM python:3.7
+FROM python:3.9
```
Comment (epicadk) [inline Dockerfile:1]: Please solve the merge conflicts.

**Review by epicadk — CHANGES_REQUESTED** [review]

**Comment by isabelcosta** [comment]
@bartekpacia if you do `poetry add psycopg2` then commit, that should work. This was failing because locally we use SQLite, and on DEV (and me on my machine) we use PostgresSQL. using locally SQLite has been very helpful and make setup easier to newcomers.

I tested it, and it worked when I tested. Let me know if you'd like to continue this work, if not someone else can continue 🙏🏾 it's 100% OK

cc @anitab-org/mentorship-maintainers

**Review by isabelcosta — CHANGES_REQUESTED** [review]
I think this would work:
```
poetry add psycopg2
```

Thank you for your patience!

I explained more here: https://github.com/anitab-org/mentorship-backend/pull/914#issuecomment-817314411

**Comment by devkapilbansal** [comment]
Closing this as #1077 is opened for this

**Pull request closed without merging**